README


Ranger 0.1.4

Written by:

  Marvin N. Wright
  Institut für Medizinische Biometrie und Statistik
  Universität zu Lübeck
  Ratzeburger Allee 160
  23562 Lübeck

  http://www.imbs-luebeck.de
  wright@imbs.uni-luebeck.de

1. Introduction

  Ranger is a fast implementation of Random Forest (Breiman 2001) or recursive partitioning, particularly suited for high dimensional data. Classification, regression, and survival forests are supported. Classification and regression forests are implemented as in the original Random Forest (Breiman 2001), survival forests as in Random Survival Forests (Ishwaran et al. 2008). 

  Ranger is written in C++, but a version for R is avaiable, too. We recommend to use the R version. It is easy to install and use and the results are readily available for further analysis. The R version is as fast as the pure C++ version.

2. Installation

  To install the Ranger R package from the provided archive just run

    install.packages("ranger_X.Y.Z.tar.gz") 

  in R from the directory where the downloaded archive is or give the full path. If any dependencies are missing (i.e. Rcpp), please install them from CRAN. R version >= 3.1 is required. Note that, for now, in Microsoft Windows no multithreading is supported (The compiler in RTools is too old).

  To install the C++ version of Ranger in Linux or Mac OS X you will need a compiler supporting C++11 (i.e. gcc >= 4.7 or Clang >= 3.0) and Cmake. To build start a terminal from the Ranger main directory and run the following commands

    cd source
    mkdir build
    cd build
    cmake ..
    make

  After compilation there should be an executable called "ranger" in the current directory. 

  To run Ranger in Microsoft Windows please use one of the provided binaries.

3. Usage

  For usage of the R version see ?ranger in R. As a first example you could try ranger(Species ~ ., data = iris).

  In the C++ version type 

    ranger --help 

  for a list of commands. First you need a training dataset in a file. This file should contain one header line with variable names and one line with variable values per sample. Variable names must not contain any whitespace, comma or semicolon. Values can be seperated by whitespace, comma or semicolon but can not be mixed in one file. A typical call of Ranger would be for example

    ranger --verbose --file data.dat --depvarname Species --treetype 1 --ntree 1000 --nthreads 4

  If you find any bugs, or if you experience any crashes, please report to us. If you have any questions just ask, we won't bite. 

References

  Breiman, L. (2001). Random forests. Machine learning, 45(1), 5-32.
  Ishwaran, H., Kogalur, U. B., Blackstone, E. H., & Lauer, M. S. (2008). Random survival forests. The Annals of Applied Statistics, 841-860.


NEWS.md
##### Version 0.5.1
* Bug fixes

##### Version 0.5.0
* New CRAN version.

##### Version 0.4.7
* Add splitting by maximally selected rank statistics for regression forests

##### Version 0.4.6
* Bug fixes

##### Version 0.4.5
* Use faster method for unordered factor splitting

##### Version 0.4.4
* Add p-values for variable importance
* Bug fixes

##### Version 0.4.3
* Add splitting by maximally selected rank statistics for survival forests
* Bug fixes

##### Version 0.4.2
* Add Windows multithreading support for new toolchain

##### Version 0.4.1
* Runtime improvement for regression forests on classification data

##### Version 0.4.0
* New CRAN version. New CRAN versions will be 0.x.0, development versions 0.x.y

##### Version 0.3.9
* Reduce memory usage of savest forest objects (changed child.nodeIDs interface)

##### Version 0.3.8
* Remove tuning functions, please use mlr or caret

##### Version 0.3.7
* Fix bug with alternative interface and prediction
* Small fixes

##### Version 0.3.6
* Add keep.inbag option to track in-bag counts
* Add option sample.fraction for fraction of sampled observations

##### Version 0.3.5
* Add tree-wise split.select.weights

##### Version 0.3.4
* Add predict.all option in predict() to get individual predictions for each tree for classification and regression
* Small changes in documentation

##### Version 0.3.3
* Add case-specific random forests

##### Version 0.3.2
* Add case weights (weighted bootstrapping or subsampling)

##### Version 0.3.1
* Catch error of outdated gcc not supporting C++11 completely

##### Version 0.3.0
* Allow the user to interrupt computation from R
* Transpose classification.table and rename to confusion.matrix
* Respect R seed for prediction
* Memory improvements for variable importance computation
* Fix bug: Probability prediction for single observations
* Fix bug: Results not identical when using alternative interface

##### Version 0.2.7 
* Small fixes for Solaris compiler

##### Version 0.2.6 
* Add C-index splitting
* Fix NA SNP handling

##### Version 0.2.5 
* Fix matrix and gwaa alternative survival interface
* Version submitted to JSS

##### Version 0.2.4 
* Small changes in documentation

##### Version 0.2.3 
* Preallocate memory for splitting

##### Version 0.2.2 
* Remove recursive splitting

##### Version 0.2.1 
* Allow matrix as input data in R version

##### Version 0.2.0 
* Fix prediction of classification forests in R

##### Version 0.1.9 
* Speedup growing for continuous covariates
* Add memory save option to save memory for very large datasets (but slower)
* Remove memory mode option from R version since no performance gain

##### Version 0.1.8 
* Fix problems when using Rcpp <0.11.4

##### Version 0.1.7 
* Add option to split on unordered categorical covariates

##### Version 0.1.6 
* Optimize memory management for very large survival forests

##### Version  0.1.5 
* Set required Rcpp version to 0.11.2
* Fix large $call objects when using BatchJobs
* Add details and example on GenABEL usage to documentation
* Minor changes to documentation

##### Version 0.1.4 
* Speedup for survival forests with continuous covariates
* R version: Generate seed from R. It is no longer necessary to set the
  seed argument in ranger calls.

##### Version 0.1.3 
* Windows support for R version (without multithreading)

##### Version 0.1.2 
* Speedup growing of regression and probability prediction forests
* Prediction forests are now handled like regression forests: MSE used for
	prediction error and permutation importance
* Fixed name conflict with randomForest package for "importance"
* Fixed a bug: prediction function is now working for probability
	prediction forests
* Slot "predictions" for probability forests now contains class probabilities
* importance function is now working even if randomForest package is
	loaded after ranger
* Fixed a bug: Split selection weights are now working as expected
* Small changes in documentation


R_package/ranger.R
# -------------------------------------------------------------------------------
#   This file is part of Ranger.
#
# Ranger is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ranger is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ranger. If not, see <http://www.gnu.org/licenses/>.
#
# Written by:
#
#   Marvin N. Wright
# Institut für Medizinische Biometrie und Statistik
# Universität zu Lübeck
# Ratzeburger Allee 160
# 23562 Lübeck
#
# http://www.imbs-luebeck.de
# wright@imbs.uni-luebeck.de
# -------------------------------------------------------------------------------

##' Ranger is a fast implementation of Random Forest (Breiman 2001) or recursive partitioning, particularly suited for high dimensional data.
##' Classification, regression, and survival forests are supported.
##' Classification and regression forests are implemented as in the original Random Forest (Breiman 2001), survival forests as in Random Survival Forests (Ishwaran et al. 2008).
##'
##' The tree type is determined by the type of the dependent variable.
##' For factors classification trees are grown, for numeric values regression trees and for survival objects survival trees.
##' The Gini index is used as splitting rule for classification, the estimated response variances for regression and the log-rank test for survival.
##' For Survival the log-rank test or an AUC-based splitting rule are available.
##'
##' With the \code{probability} option and factor dependent variable a probability forest is grown.
##' Here, the estimated response variances are used for splitting, as in regression forests.
##' Predictions are class probabilities for each sample.
##' For details see Malley et al. (2012).
##'
##' Note that for classification and regression nodes with size smaller than min.node.size can occur, like in original Random Forest.
##' For survival all nodes contain at least min.node.size samples. 
##' Variables selected with \code{always.split.variables} are tried additionaly to the mtry variables randomly selected.
##' In \code{split.select.weights} variables weighted with 0 are never selected and variables with 1 are always selected. 
##' Weights do not need to sum up to 1, they will be normalized later. 
##' The usage of \code{split.select.weights} can increase the computation times for large forests.
##'
##' For a large number of variables and data frame as input data the formula interface can be slow.
##' Alternatively dependent.variable.name (and status.variable.name for survival) can be used.
##' By setting \code{memory} to 'float' or 'char' the memory usage can be reduced with a possible loss of precision. 
##' Use char only if all endpoints and covariates are integer values between -128 and 127 or factors with few levels. 
##' 
##' For GWAS data consider combining \code{ranger} with the \code{\link{GenABEL}} package. 
##' See the Examples section below for a demonstration using \code{Plink} data.
##' All SNPs in the \code{GenABEL} object will be used for splitting. 
##' To use only the SNPs without sex or other covariates from the phenotype file, use \code{0} on the right hand side of the formula. 
##' Note that missing values are treated as an extra category while splitting.
##' 
##' Notes:
##' \itemize{
##'  \item Multithreading is currently not supported for Microsoft Windows platforms.
##' }
##' 
##' @title Ranger
##' @param formula Object of class \code{formula} or \code{character} describing the model to fit.
##' @param data Training data of class \code{data.frame} or \code{gwaa.data} (GenABEL).
##' @param num.trees Number of trees.
##' @param mtry Number of variables to possibly split at in each node.
##' @param importance Variable importance mode, one of 'none', 'impurity', 'permutation'. The 'impurity' measure is the Gini index for classification and the variance of the responses for regression.
##' @param write.forest Save \code{ranger.forest} object, needed for prediction.
##' @param probability Grow a probability forest. This is a classification forest which returns class probabilities instead of classifications.
##' @param min.node.size Minimal node size. Default 1 for classification, 5 for regression, 3 for survival, and 10 for probability.
##' @param replace Sample with replacement. Default TRUE.
##' @param splitrule Splitting rule, survival only. The splitting rule can be chosen of "logrank", "auc" and "auc_ignore_ties" with default "logrank". 
##' @param split.select.weights Numeric vector with weights between 0 and 1, representing the probability to select variables for splitting.  
##' @param always.split.variables Character vector with variable names to be always tried for splitting.
##' @param respect.unordered.factors Regard unordered factor covariates as unordered categorical variables. If \code{FALSE}, all factors are regarded ordered. 
##' @param scale.permutation.importance Scale permutation importance by standard error as in (Breiman 2001). Only applicable if permutation variable importance mode selected.
##' @param num.threads Number of threads. Default is number of CPUs available.
##' @param verbose Verbose output on or off.
##' @param seed Random seed. Default is \code{NULL}, which generates the seed from \code{R}. 
##' @param memory Memory mode, one of 'double', 'float', 'char'. Default 'double'.
##' @param dependent.variable.name Name of dependent variable, needed if no formula given. For survival forests this is the time variable.
##' @param status.variable.name Name of status variable, only applicable to survival data and needed if no formula given. Use 1 for event and 0 for censoring.
##' @return Object of class \code{ranger} with elements
##'   \tabular{ll}{
##'       \code{forest} \tab Saved forest (If write.forest set to TRUE). Note that the variable IDs in the \code{split.varIDs} object do not necessarily represent the column number in R. \cr
##'       \code{predictions}    \tab Predicted classes/values, based on out of bag samples (classification and regression only). \cr
##'       \code{variable.importance}     \tab Variable importance for each independent variable. \cr
##'       \code{prediction.error}   \tab Overall out of bag prediction error. For classification this is the fraction of missclassified samples, for regression the mean squared error and for survival one minus Harrell's c-index. \cr
##'       \code{r.squared}   \tab R squared. Also called explained variance or coefficient of determination (regression only). \cr
##'       \code{classification.table} \tab Contingency table for classes and predictions based on out of bag samples (classification only). \cr
##'       \code{unique.death.times} \tab Unique death times (survival only). \cr
##'       \code{chf} \tab Estimated cumulative hazard function for each sample (survival only). \cr
##'       \code{survival} \tab Estimated survival function for each sample (survival only). \cr
##'       \code{call}    \tab Function call. \cr
##'       \code{num.trees}   \tab Number of trees. \cr
##'       \code{num.independent.variables} \tab Number of independent variables. \cr
##'       \code{mtry}    \tab Value of mtry used. \cr
##'       \code{min.node.size}   \tab Value of minimal node size used. \cr
##'       \code{treetype}    \tab Type of forest/tree. classification, regression or survival. \cr
##'       \code{memory.mode}   \tab Memory mode used. \cr
##'       \code{importance.mode}     \tab Importance mode used. \cr
##'       \code{num.samples}     \tab Number of samples.
##'   }
##' @examples
##' require(ranger)
##'
##' ## Classification forest with default settings
##' ranger(Species ~ ., data = iris)
##'
##' ## Prediction
##' train.idx <- sample(nrow(iris), 2/3 * nrow(iris))
##' iris.train <- iris[train.idx, ]
##' iris.test <- iris[-train.idx, ]
##' rg.iris <- ranger(Species ~ ., data = iris.train, write.forest = TRUE)
##' pred.iris <- predict(rg.iris, dat = iris.test)
##' table(iris.test$Species, pred.iris$predictions)
##'
##' ## Variable importance
##' rg.iris <- ranger(Species ~ ., data = iris, importance = "impurity")
##' rg.iris$variable.importance
##'
##' ## Survival forest
##' require(survival)
##' rg.veteran <- ranger(Surv(time, status) ~ ., data = veteran)
##' plot(rg.veteran$unique.death.times, rg.veteran$survival[1,])
##'
##' ## Alternative interface
##' ranger(dependent.variable.name = "Species", data = iris)
##' 
##' \dontrun{
##' ## Use GenABEL interface to read Plink data into R and grow a classification forest
##' ## The ped and map files are not included
##' library(GenABEL)
##' convert.snp.ped("data.ped", "data.map", "data.raw")
##' dat.gwaa <- load.gwaa.data("data.pheno", "data.raw")
##' phdata(dat.gwaa)$trait <- factor(phdata(dat.gwaa)$trait)
##' ranger(trait ~ ., data = dat.gwaa)
##' }
##'
##' @author Marvin N. Wright
##' @references
##'   Breiman, L. (2001). Random forests. Mach Learn, 45(1), 5-32. \cr
##'   Ishwaran, H., Kogalur, U. B., Blackstone, E. H., & Lauer, M. S. (2008). Random survival forests. Ann Appl Stat, 841-860. \cr
##'   Malley, J. D., Kruppa, J., Dasgupta, A., Malley, K. G., & Ziegler, A. (2012). Probability machines: consistent probability estimation using nonparametric learning machines. Methods Inf Med, 51(1), 74.
##' @seealso \code{\link{predict.ranger}}
##' @export
ranger <- function(formula = NULL, data = NULL, num.trees = 500, mtry = NULL,
                   importance = "none", write.forest = FALSE, probability = FALSE,
                   min.node.size = NULL, replace = TRUE, splitrule = NULL,
                   split.select.weights = NULL, always.split.variables = NULL,
                   respect.unordered.factors = FALSE,
                   scale.permutation.importance = FALSE,
                   num.threads = NULL,
                   verbose = TRUE, seed = NULL, memory = "double",
                   dependent.variable.name = NULL, status.variable.name = NULL) {
  
  ## GenABEL GWA data
  if (class(data) == "gwaa.data") {
    snp.names <- data@gtdata@snpnames
    sparse.data <- data@gtdata@gtps@.Data
    data <- data@phdata
    if ("id" %in% names(data)) {
      data$"id" <- NULL
    }
    gwa.mode <- TRUE
  } else {
    sparse.data <- as.matrix(0)
    gwa.mode <- FALSE
  }
  
  ## Formula interface. Use whole data frame is no formula provided and depvarname given
  if (is.null(formula)) {
    if (is.null(dependent.variable.name)) {
      stop("Error: Please give formula or dependent variable name.")
    }
    if (is.null(status.variable.name)) {
      status.variable.name <- "none"
      response <- data[, dependent.variable.name]
    } else {
      response <- data[, c(dependent.variable.name, status.variable.name)]
    }
    data.selected <- data
  } else {
    formula <- formula(formula)
    if (class(formula) != "formula") {
      stop("Error: Invalid formula.")
    }
    data.selected <- model.frame(formula, data, na.action = na.fail)
    response <- data.selected[[1]]
  }
  
  ## Probability estimation
  if (probability & !is.factor(response)) {
    stop("Error: Probability estimation is only applicable to categorical (factor) dependent variables.")
  }
  
  ## Treetype
  if (is.factor(response)) {
    if (probability) {
      treetype <- 9
    } else {
      treetype <- 1
    }
  } else if (is.numeric(response) & is.vector(response)) {
    treetype <- 3
  } else if (class(response) == "Surv" | class(response) == "data.frame") {
    treetype <- 5
  } else {
    stop("Error: Unsupported type of dependent variable.")
  }
  
  ## Dependent and status variable name. For non-survival dummy status variable name.
  if (!is.null(formula)) {
    if (treetype == 5) {
      dependent.variable.name <- dimnames(response)[[2]][1]
      status.variable.name <- dimnames(response)[[2]][2]
    } else {
      dependent.variable.name <- names(data.selected)[1]
      status.variable.name <- "none"
    }
    independent.variable.names <- names(data.selected)[-1]
  } else {
    independent.variable.names <- names(data.selected)[names(data.selected) != dependent.variable.name &
                                                         names(data.selected) != status.variable.name]
  }
  
  ## Input data and variable names
  if (!is.null(formula)) {
    if (treetype == 5) {
      data.final <- data.matrix(cbind(response[, 1], response[, 2],
                                      data.selected[-1]))
      variable.names <- c(dependent.variable.name, status.variable.name,
                          independent.variable.names)
    } else {
      data.final <- data.matrix(data.selected)
      variable.names <- names(data.selected)
    }
  } else {
    data.final <- data.matrix(data.selected)
    variable.names <- names(data.selected)
  }
  
  ## If gwa mode, add snp variable names
  if (gwa.mode) {
    variable.names <- c(variable.names, snp.names)
    all.independent.variable.names <- c(independent.variable.names, snp.names)
  } else {
    all.independent.variable.names <- independent.variable.names
  }
  
  ## Number of trees
  if (!is.numeric(num.trees) | num.trees < 1) {
    stop("Error: Invalid value for num.trees.")
  }
  
  ## mtry
  if (is.null(mtry)) {
    mtry <- 0
  } else if (!is.numeric(mtry) | mtry < 0) {
    stop("Error: Invalid value for mtry")
  }
  
  ## Memory mode
  if (is.null(memory) | memory == "double") {
    memory.mode <- 0
  } else if (memory == "float") {
    memory.mode <- 1
  } else if (memory == "char") {
    memory.mode <- 2
  } else {
    stop("Error: Unknown memory mode.")
  }
  
  ## Seed
  if (is.null(seed)) {
    seed <- runif(1 , 0, .Machine$integer.max)
  }
  
  ## Num threads
  ## Default 0 -> detect from system in C++.
  if (is.null(num.threads)) {
    num.threads = 0
  } else if (!is.numeric(num.threads) | num.threads < 0) {
    stop("Error: Invalid value for num.threads")
  }
  
  ## Minumum node size
  if (is.null(min.node.size)) {
    min.node.size <- 0
  } else if (!is.numeric(min.node.size) | min.node.size < 0) {
    stop("Error: Invalid value for min.node.size")
  }
  
  ## Importance mode
  if (is.null(importance) | importance == "none") {
    importance.mode <- 0
  } else if (importance == "impurity") {
    importance.mode <- 1
    if (treetype == 5) {
      stop("Node impurity variable importance not supported for survival forests.")
    }
  } else if (importance == "permutation") {
    if (scale.permutation.importance) {
      importance.mode <- 2
    } else {
      importance.mode <- 3
    }
  } else {
    stop("Error: Unknown importance mode.")
  }
  
  ## Split select weights: NULL for no weights
  if (is.null(split.select.weights)) {
    split.select.weights <- c(0,0)
    use.split.select.weights <- FALSE
  } else {
    use.split.select.weights <- TRUE
  }
  
  ## Always split variables: NULL for no variables
  if (is.null(always.split.variables)) {
    always.split.variables <- c("0", "0")
    use.always.split.variables <- FALSE
  } else {
    use.always.split.variables <- TRUE
  }
  
  if (use.split.select.weights & use.always.split.variables) {
    stop("Error: Please use only one option of use.split.select.weights and use.always.split.variables.")
  }
  
  ## Splitting rule
  if (is.null(splitrule)) {
    splitrule <- 1
  } else if (treetype == 5 & splitrule == "logrank") {
    splitrule <- 1
  } else if (treetype == 5 & splitrule == "auc") {
    splitrule <- 2
  } else if (treetype == 5 & splitrule == "auc_ignore_ties") {
    splitrule <- 3
  } else {
    stop("Error: Unknown splitrule.")
  }

  ## Unordered factors  
  if (respect.unordered.factors) {
    names.selected <- names(data.selected)
    ordered.idx <- sapply(data.selected, is.ordered)
    factor.idx <- sapply(data.selected, is.factor)
    independent.idx <- names.selected != dependent.variable.name & names.selected != status.variable.name
    unordered.factor.variables <- names.selected[factor.idx & !ordered.idx & independent.idx]
    
    if (length(unordered.factor.variables) > 0) {
      use.unordered.factor.variables <- TRUE
      ## Check level count
      num.levels <- sapply(data.selected[, factor.idx & !ordered.idx & independent.idx, drop = FALSE], nlevels)
      max.level.count <- 8*.Machine$sizeof.pointer - 1
      if (max(num.levels) > max.level.count) {
        stop(paste("Too many levels in unordered categorical variable ", unordered.factor.variables[which.max(num.levels)], 
                   ". Only ", max.level.count, " levels allowed on this system. Consider ordering this factor.", sep = ""))
      } 
    } else {
      unordered.factor.variables <- c("0", "0")
      use.unordered.factor.variables <- FALSE
    } 
  } else {
    unordered.factor.variables <- c("0", "0")
    use.unordered.factor.variables <- FALSE
  }
  
  ## Prediction mode always false. Use predict.ranger() method.
  prediction.mode <- FALSE
  
  ## No loaded forest object
  loaded.forest <- list()
  
  ## Call Ranger
  result <- rangerCpp(treetype, dependent.variable.name, memory.mode, data.final, variable.names, mtry,
                      num.trees, verbose, seed, num.threads, write.forest, importance.mode,
                      min.node.size, split.select.weights, use.split.select.weights,
                      always.split.variables, use.always.split.variables,
                      status.variable.name, prediction.mode, loaded.forest, sparse.data,
                      replace, probability, unordered.factor.variables, use.unordered.factor.variables, 
                      splitrule)
  
  if (length(result) == 0) {
    stop("Internal error.")
  }
  
  ## Prepare results
  result$predictions <- drop(do.call(rbind, result$predictions))
  if (importance.mode != 0) {
    names(result$variable.importance) <- all.independent.variable.names
  }
  
  ## Set predictions
  if (treetype == 1) {
    result$predictions <- factor(result$predictions, levels = 1:nlevels(response),
                                 labels = levels(response))
    result$classification.table <- table(result$predictions, unlist(data[, dependent.variable.name]), dnn = c("predicted", "true"))
  } else if (treetype == 5) {
    result$chf <- result$predictions
    result$predictions <- NULL
    result$survival <- exp(-result$chf)
  } else if (treetype == 9) {
    colnames(result$predictions) <- levels(response)
  }
  
  ## Set treetype
  if (treetype == 1) {
    result$treetype <- "Classification"
  } else if (treetype == 3) {
    result$treetype <- "Regression"
  } else if (treetype == 5) {
    result$treetype <- "Survival"
  } else if (treetype == 9) {
    result$treetype <- "Probability estimation"
  }
  if (treetype == 3) {
    result$r.squared <- 1 - result$prediction.error / var(response)
  }
  result$call <- sys.call()
  result$memory.mode <- memory
  result$importance.mode <- importance
  result$num.samples <- nrow(data.final)
  
  ## Write forest object
  if (write.forest) {
    result$forest$levels <- levels(response)
    result$forest$independent.variable.names <- independent.variable.names
    result$forest$treetype <- result$treetype
    class(result$forest) <- "ranger.forest"
  }
  
  class(result) <- "ranger"
  return(result)
}




